src/apps/app_bulk_upload.erl

```diff
@@ -0,0 +1,173 @@
+upload(Node, Wallet, Blob) ->
+	Queue = app_queue:start(Node, Wallet),
+	Hash = crypto:hash(?BLOB_HASH_ALGO, Blob),
+	upload_blob(Queue, Hash, Blob),
+	Queue.
+
+%% @doc Takes a binary blob and processes it chunk by chunk. Each chunk is converted into
+%% a transaction and put into the queue. Chunk size is 1MB.
+upload_blob(Queue, Hash, Blob) ->
+	BlobSize = byte_size(Blob),
+	ChunkNumber = BlobSize div ?MB + if BlobSize rem ?MB =/= 0 -> 1; true -> 0 end,
+	upload_blob(Queue, Hash, Blob, ChunkNumber, 1).
+
+upload_blob(Queue, Hash, Blob, ChunkNumber, ChunkPosition) ->
+	if byte_size(Blob) =< ?MB ->
+		app_queue:add(Queue, chunk_to_tx(Hash, Blob, ChunkNumber, ChunkPosition));
+	true ->
+		<< Chunk:?MB/binary, Rest/binary >> = Blob,
+		app_queue:add(Queue, chunk_to_tx(Hash, Chunk, ChunkNumber, ChunkPosition)),
+		upload_blob(Queue, Hash, Rest, ChunkNumber, ChunkPosition + 1)
+	end.
+
+%% @doc Converts the given binary chunk into a transaction. A hash of the whole block the chunk
+%% is part of is assigned as a tag.
+chunk_to_tx(Hash, Chunk, ChunkNumber, ChunkPosition) ->
+	#tx {
+		tags =
+			[
+				{"app_name", "BulkUpload"},
+				{"blob_hash", ar_util:encode(Hash)},
+				{"number_of_chunks", integer_to_binary(ChunkNumber)},
+				{"chunk_position", integer_to_binary(ChunkPosition)}
+			],
+		data = Chunk
+	}.
+
+%% @doc Searches the local storage for the chunks of the blob with the given hash.
+%% If the blob is reconstructed successfuly, writes it to the specified destination.
+%% The provided hash has to be a Base 64 encoded SHA 256 hash of the file as a binary string.
+download(Hash, Filename) ->
+	{ok, Blob} = download(Hash),
+	file:write_file(Filename, Blob, [{encoding, unicode}]).
+
+download(Hash) ->
+	app_search:get_entries(<< "blob_hash" >>, Hash),
+	receive TXIDs ->
+		Transactions = lists:map(fun(TX) -> ar_storage:read_tx(TX) end, TXIDs),
+		{ok, Blob} = reconstruct_blob(Transactions),
+		BlobHash = ar_util:encode(crypto:hash(?BLOB_HASH_ALGO, Blob)),
+		if BlobHash /= Hash ->
+			{error, invalid_upload};
+		true ->
+			{ok, Blob}
+		end
+	end.
+
+reconstruct_blob([]) -> {error, not_found};
+reconstruct_blob(Transactions) ->
+	Head = hd(Transactions),
+	{_, ChunkNumberBinary} = lists:keyfind(<< "number_of_chunks" >>, 1, Head#tx.tags),
+	ChunkNumber = binary_to_integer(ChunkNumberBinary),
+	if length(Transactions) < ChunkNumber ->
+		{error, missing_chunks};
+	true ->
+		SortedTransactions = lists:sort(fun(First, Second) ->
+			{_, FirstPositionBinary} = lists:keyfind(<< "chunk_position" >>, 1, First#tx.tags),
+			{_, SecondPositionBinary} = lists:keyfind(<< "chunk_position" >>, 1, Second#tx.tags),
+			FirstChunkPosition = binary_to_integer(FirstPositionBinary),
+			SecondChunkPosition = binary_to_integer(SecondPositionBinary),
+			FirstChunkPosition =< SecondChunkPosition
+		end,
+		Transactions
+		),
+		{ok, << (TX#tx.data) || TX <- SortedTransactions >>}
+	end.
+
+
+upload_test_() ->
+	{timeout, 60, fun() ->
+		app_search:deleteDB(),
```

Comment:
Ah, so you mean that the test will fail the second time since the `blob_hash` will collide. Wait a minute.... That will happen in reality too... I will add more comments about this.

Reply:
Ah yes, the duplicate transactions may create a problem!

> I will add more comments about this.

Let's just discuss it in this thread for convenience.